### PR TITLE
[NEXUS-8619] apply retry logic to raw put

### PR DESCRIPTION
**Pull request by mcculls**
https://issues.sonatype.org/browse/NEXUS-8619
http://bamboo.s/browse/NX3-OSSF406-2 :white_check_mark:

### CODE REVIEW

**Comment by adreghiciu** [comment]
+1

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageFacetImpl.java [hunk 1]
```diff
@@ -245,15 +247,15 @@ public void unregisterHookSupplier(final Supplier<StorageTxHook> hookSupplier) {
           tx.commit();
           return result;
         }
-        catch (OConcurrentModificationException e) {
-          lastConcurrentException = e;
+        catch (OConcurrentModificationException | ORecordDuplicatedException e) {
```
Comment (cstamas) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageFacetImpl.java]: Am a bit afraid about ORecordDuplicatedException... this implies that Operation must be Create-Or-Update aware, like I believe most are...

**Comment by cstamas** [comment]
+1 but this change might mean we can remove this try-catch completely (and leave the `store(context, remote);` like unguarded).
https://github.com/sonatype/nexus-oss/blob/master/components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/ProxyFacetSupport.java#L158

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageFacetImpl.java (hunk 1 above)
Comment (mcculls) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageFacetImpl.java]: The places which currently use the `perform` method are all create-or-update - if we don't retry on this exception then you can get failures when two threads attempt to create the same logical entity at exactly the same time.

When we switch to the method aspect solution we'll handle create-only cases differently.

**Comment by mcculls** [comment]
Agreed, removed check in 79f0664 since this is now handled in `StorageFacetImpl.perform`

**Comment by mcculls** [comment]
http://bamboo.s/browse/NX3-OSSF406-3 :white_check_mark:

**Comment by mrprescott** [comment]
+1

**Comment by mcculls** [comment]
Merged to master as 6054e04

**Pull request closed without merging**